test/common/router/config_impl_speed_test.cc

```diff
@@ -132,13 +142,260 @@ static void bmRouteTableSizeWithExactPathMatch(benchmark::State& state) {
+}
+
+// Arg1 is the path rule table size. Arg2 is the path rule length.
+BENCHMARK(bmRouteTableSizeWithExactPathMatchVaryPathLen)
+    ->ArgPair(1, 10)
+    ->ArgPair(10, 10)
+    ->ArgPair(50, 10)
+    ->ArgPair(100, 10)
+    ->ArgPair(500, 10)
+    ->ArgPair(1000, 10)
+    ->ArgPair(1500, 10)
+    ->ArgPair(2000, 10)
+    ->ArgPair(5000, 10)
+    ->ArgPair(10000, 10)
+    ->ArgPair(1, 50)
+    ->ArgPair(10, 50)
+    ->ArgPair(50, 50)
+    ->ArgPair(100, 50)
+    ->ArgPair(500, 50)
+    ->ArgPair(1000, 50)
+    ->ArgPair(1500, 50)
+    ->ArgPair(2000, 50)
+    ->ArgPair(5000, 50)
+    ->ArgPair(10000, 50)
+    ->ArgPair(1, 100)
+    ->ArgPair(10, 100)
+    ->ArgPair(50, 100)
+    ->ArgPair(100, 100)
+    ->ArgPair(500, 100)
+    ->ArgPair(1000, 100)
+    ->ArgPair(1500, 100)
+    ->ArgPair(2000, 100)
+    ->ArgPair(5000, 100)
+    ->ArgPair(10000, 100)
+    ->ArgPair(1, 500)
+    ->ArgPair(10, 500)
+    ->ArgPair(50, 500)
+    ->ArgPair(100, 500)
+    ->ArgPair(500, 500)
+    ->ArgPair(1000, 500)
+    ->ArgPair(1500, 500)
+    ->ArgPair(2000, 500)
+    ->ArgPair(5000, 500)
+    ->ArgPair(10000, 500)
+    ->ArgPair(1, 1000)
+    ->ArgPair(10, 1000)
+    ->ArgPair(50, 1000)
+    ->ArgPair(100, 1000)
+    ->ArgPair(500, 1000)
+    ->ArgPair(1000, 1000)
+    ->ArgPair(1500, 1000)
+    ->ArgPair(2000, 1000)
+    ->ArgPair(5000, 1000)
+    ->ArgPair(10000, 1000)
+    ->ArgPair(1, 2000)
+    ->ArgPair(10, 2000)
+    ->ArgPair(50, 2000)
+    ->ArgPair(100, 2000)
+    ->ArgPair(500, 2000)
+    ->ArgPair(1000, 2000)
+    ->ArgPair(1500, 2000)
+    ->ArgPair(2000, 2000)
+    ->ArgPair(5000, 2000)
+    ->ArgPair(10000, 2000);
+
+/**
+ * Benchmark a route table with regex path matchers in the form of:
+ * - "^/api/_?/v\\d{2}"                                       : ~10 bytes regex complexity
+ * - "^/videos/_?/v\\d{3}/ba[r|z].m?[4s|pg|p4]"               " ~50 bytes regex complexity
+ * - "^/.* /foo\\d{3}\\W{1,2}V+T?/ba[r|z]/A{2}.[h|w][^a-z]{1}": ~100 bytes regex complexity
+ * - etc.
+ */
+
+// Generate RegEx path rule based on path rule length.
+static std::string GetRegexPathRuleFromBase(int path_len) {
+  std::string path_rule = "";
+
+  switch (path_len) {
+    case 10:
+      absl::StrAppend(&path_rule, "^/api/_?/v\\d{2}");
+      break;
+    case 50:
+      absl::StrAppend(&path_rule, "^/videos/_?/v\\d{3}/ba[r|z].m?[4s|pg|p4]");
+      break;
+    case 100:
+      absl::StrAppend(&path_rule,
+            "^/.*/foo\\d{3}\\W{1,2}V+T?/ba[r|z]/A{2}.[h|w][^a-z]{1}");
+      break;
+  }
+  return path_rule;
+}
+
+// Generate request header to match corresponding RexEx rule.
+static std::string GetRegexPathHeaderFromBase(int path_len) {
+  std::string path_header = "";
+
+  switch (path_len) {
+    case 10:
+      absl::StrAppend(&path_header, "/api/_/v12");
+      break;
+    case 50:
+      absl::StrAppend(&path_header, "/videos/_/v123/bar.mp4");
+      break;
+    case 100:
+      absl::StrAppend(&path_header,
+            "/test/foo123&&VT/baz/AA.wB");
```

Comment:
I'm not sure how to think about the 50 vs. 100 case here. They don't seem to be double the length/complexity?

Reply:
I can not find a nice online tool to calculate the regex complexity. Thus what I am doing is to utilize Envoy error checking as below during booting up with certain envoy configs:
[2021-06-08 19:55:44.636][1836312][info][main] [source/server/server.cc:494] admin address: 0.0.0.0:9901
[2021-06-08 19:55:44.637][1836312][info][main] [source/server/server.cc:659] runtime: layers:
  - name: base
    static_layer:
      {}
  - name: admin

    admin_layer:
      {}
[2021-06-08 19:55:44.637][1836312][info][config] [source/server/configuration_impl.cc:128] loading tracing configuration
[2021-06-08 19:55:44.637][1836312][info][config] [source/server/configuration_impl.cc:88] loading 0 static secret(s)
[2021-06-08 19:55:44.637][1836312][info][config] [source/server/configuration_impl.cc:94] loading 1 cluster(s)
[2021-06-08 19:55:44.638][1836312][info][config] [source/server/configuration_impl.cc:98] loading 1 listener(s)
[2021-06-08 19:55:44.642][1836312][critical][main] [source/server/server.cc:109] error initializing configuration 'configs/envoy-regex-count-1-predicate-100.yaml': regex '^/.*/foo\\d{5}\\W{1,2}V+T?/ba[r|z]/A{4}.[h|w][^a-z]{6}0' RE2 program size of 101 > max program size of 100 set for the error level threshold. Increase configured max program size if necessary.
[2021-06-08 19:55:44.642][1836312][info][main] [source/server/server.cc:807] exiting
regex '^/.*/foo\\d{5}\\W{1,2}V+T?/ba[r|z]/A{4}.[h|w][^a-z]{6}0' RE2 program size of 101 > max program size of 100 set for the error level threshold. Increase configured max program size if necessary.


So, with above regex path config: '^/.*/foo\\d{5}\\W{1,2}V+T?/ba[r|z]/A{4}.[h|w][^a-z]{6}0', the RE2 program size is 101, which is exceeding the default max_program_size: 100.  Then I removed the trailing "0" from the regex, and see the Envoy can boot up fine with that config.  BTW, adding another "0" after the above regex will show RE2 program size is 102.

For 50 complexity case, I basically duplicate a regex to exceed the 100 limit to trigger the warning.  Then removing some characters.  

This probably is not the best way to get the regex complexity size.

Reply:
Can you do something more foundational, e.g. figure out how much a `.*` or other regex feature contributes towards complexity?

Reply:
.* contributes 8 bytes to the regex complexity based on my test.

Reply:
OK, then put in 6 or 12 of them :)